pynumaflow/batchmapper/_dtypes.py

```diff
@@ -0,0 +1,258 @@
+        self._watermark = watermark
+        self._headers = headers or {}
+
+    @property
+    def id(self) -> str:
+        """Returns the keys of the event"""
+        return self._id
+
+    @property
+    def keys(self) -> list[str]:
+        """Returns the keys of the event"""
+        return self._keys
+
+    @property
+    def value(self) -> bytes:
+        """Returns the value of the event."""
+        return self._value
+
+    @property
+    def event_time(self) -> datetime:
+        """Returns the event time of the event."""
+        return self._event_time
+
+    @property
+    def watermark(self) -> datetime:
+        """Returns the watermark of the event."""
+        return self._watermark
+
+    @property
+    def headers(self) -> dict[str, str]:
+        """Returns the headers of the event."""
+        return self._headers.copy()
+
+
+class BatchResponses:
+    # ID of input datum messages derived from
+
+    def __init__(self, id: str, messages: Messages = None):
+        self._id = id
+        self._messages = messages
+
+    @classmethod
+    def to_drop(cls: type[M], msg_id: str) -> M:
+        msgs = Messages(Message.to_drop())
+        return cls(msg_id, msgs)
+
+    @property
+    def id(self) -> str:
+        """Returns the keys of the event"""
+        return self._id
+
+    @property
+    def messages(self) -> Messages:
+        """Returns the keys of the event"""
+        return self._messages
+
+    def __repr__(self) -> str:
+        return f"BaseResponses: id={self.__class__.__name__}, num_messages={len(self._messages)}"
+
+
+class BatchMapper(metaclass=ABCMeta):
+    """
+    Provides an interface to write a SyncMapServicer
+    which will be exposed over a Synchronous gRPC server.
+    """
+
+    def __call__(self, *args, **kwargs):
+        """
+        This allows to execute the handler function directly if
+        class instance is sent as a callable.
+        """
+        return self.handler(*args, **kwargs)
+
+    @abstractmethod
+    def handler(self, datums: AsyncIterable[Datum]) -> AsyncIterable[BatchResponses]:
+        """
+        Implement this handler function which implements the MapSyncCallable interface.
+        """
+        pass
+
+
+class BatchMapperUnary(metaclass=ABCMeta):
+    """
+    Provides an interface to write a BatchMapUnaryServicer
+    which will be exposed over a Asynchronous gRPC server.
+    """
+
+    def __call__(self, *args, **kwargs):
+        """
+        This allows to execute the handler function directly if
+        class instance is sent as a callable.
+        """
+        return self.handler(*args, **kwargs)
+
+    @abstractmethod
+    def handler(self, keys: list[str], datum: Datum) -> Messages:
+        """
+        Implement this handler function which implements the MapSyncCallable interface.
+        """
+        pass
+
+
+MapBatchAsyncHandlerCallable = Callable[[AsyncIterable[Datum]], AsyncIterable[BatchResponses]]
```

Comment:
We planned on returning BatchResponses as a whole from the user code instead of streaming the results. 
So the user should be creating an array of BatchResponse ([]BatchResponse -> BatchResponses) and that should be returned. 
Where len(BatchResponses) == len(requests)